src/python/pants/base/hash_utils.py

```diff
@@ -39,6 +41,41 @@ def hash_file(path, digest=None):
   return digest.hexdigest() if PY3 else digest.hexdigest().decode('utf-8')
 
 
+class CoercingEncoder(json.JSONEncoder):
+  def default(self, o):
+    if isinstance(o, collections.Iterable) and not isinstance(o, collections.Mapping):
+      # For things like sets which JSONEncoder doesn't handle, and raises a TypeError on.
+      return list(o)
+    return super(CoercingEncoder, self).default(o)
+
+  def encode(self, o):
```

Comment:
So the trick here turns out to be to do all the work in default and just override encode with:
```python
def encode(self, o):
    return super(CoercingEncoder, self).encode(self.default(o))
```

Then you get recursive handling. Proof of concept:
```python
$ ./pants -q repl 3rdparty/python:future

Python 2.7.15 (default, Jun 27 2018, 13:05:28) 
[GCC 8.1.1 20180531] on linux2
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import json
>>> 
>>> from builtins import bytes, str
>>> from future.moves import collections
>>> 
>>> 
>>> class Encoder(json.JSONEncoder):
...   def default(self, o):
...     if isinstance(o, collections.Mapping):
...       return {self.encode(k): self.default(v) for k, v in o.items()}
...     elif isinstance(o, collections.Set):
...       return sorted(self.default(i) for i in o)
...     elif isinstance(o, collections.Iterable) and not isinstance(o, (bytes, list, str)):
...       return list(self.default(i) for i in o)
...     return o
...   def encode(self, o):
...     return super(Encoder, self).encode(self.default(o))
... 
>>> 
>>> def encode(obj, **kwargs):
...   print(json.dumps(obj, ensure_ascii=True, allow_nan=False, sort_keys=True, cls=Encoder, **kwargs))
... 
>>> encode({
...   (2, 'a'): ('a', 3, 'c'),
...   (1, 'a'): {'a', 3, 'c'},
... }, indent=2)
{
  "[\n  1, \n  \"a\"\n]": [
    3, 
    "a", 
    "c"
  ], 
  "[\n  2, \n  \"a\"\n]": [
    "a", 
    3, 
    "c"
  ]
}
>>> 
>>> encode({
...   ('a', 1): ({(42, 'g'): "bob"}, 4)
... }, indent=2)
{
  "[\n  \"a\", \n  1\n]": [
    {
      "[\n  42, \n  \"g\"\n]": "bob"
    }, 
    4
  ]
}
>>> 
>>> encode({
...   ('a', 1): (
...     {(42, 'g'): {"bob": (3, 5)}}, 4
...   )
... }, indent=2)
{
  "[\n  \"a\", \n  1\n]": [
    {
      "[\n  42, \n  \"g\"\n]": {
        "\"bob\"": [
          3, 
          5
        ]
      }
    }, 
    4
  ]
}
>>> 
```

Reply:
Thank you so much for this thorough investigation! I will implement this immediately.

Reply:
Done in 8999818c1! Thanks again for figuring this out and describing the solution so clearly!